curvefs/src/mds/space/volume_space.cpp

```diff
@@ -0,0 +1,395 @@
+        if (it != allocatedGroups_.end()) {
+            if (it->second.owner() != group.owner()) {
+                LOG(WARNING)
+                    << "Owner is not identical, block group may "
+                       "assign to others, fsId: "
+                    << fsId_ << ", block group offset: " << group.offset();
+                return SpaceErrConflict;
+            }
+
+            // space is total available
+            if (group.available() == group.size()) {
+                auto err = ClearBlockGroup(group);
+                if (err != SpaceOk) {
+                    LOG(WARNING) << "Clear block group failed, fsId: " << fsId_
+                                 << ", block group offset: " << group.offset();
+                    return err;
+                }
+
+                cleanGroups_.insert(group.offset());
+            } else {
+                auto copy = group;
+                copy.clear_owner();
+                auto err = PersistBlockGroup(copy);
+                if (err != SpaceOk) {
+                    LOG(WARNING)
+                        << "Persist block group failed, fsId: " << fsId_
+                        << ", block group offset: " << group.offset()
+                        << ", err: " << SpaceErrCode_Name(err);
+                    return err;
+                }
+
+                availableGroups_.emplace(group.offset(), std::move(copy));
+            }
+
+            allocatedGroups_.erase(group.offset());
+        }
+
+        // and if it's not allocated, this request must be a retry request
+    }
+
+    return SpaceOk;
+}
+
+SpaceErrCode VolumeSpace::PersistBlockGroup(const BlockGroup& group) {
+    return storage_->PutBlockGroup(fsId_, group.offset(), group);
+}
+
+SpaceErrCode VolumeSpace::PersistBlockGroups(
+    const std::vector<BlockGroup>& blockGroups) {
+    SpaceErrCode err = SpaceOk;
+    for (auto& group : blockGroups) {
+        err = PersistBlockGroup(group);
+
+        // TODO(wuhanqing): handle error, and rollback if necessary
+        if (err != SpaceOk) {
+            LOG(ERROR) << "Put block group failed, fsId: " << fsId_
+                       << ", offset: " << group.offset()
+                       << ", err: " << SpaceErrCode_Name(err);
+            return err;
+        }
+    }
+
+    return err;
+}
+
+SpaceErrCode VolumeSpace::ClearBlockGroup(const BlockGroup& group) {
+    auto err = storage_->RemoveBlockGroup(fsId_, group.offset());
+    if (err != SpaceOk) {
+        LOG(WARNING) << "Remove block group failed, fsId: " << fsId_
+                     << ", offset: " << group.offset()
+                     << ", err: " << SpaceErrCode_Name(err);
+    }
+
+    return err;
+}
+
+SpaceErrCode VolumeSpace::RemoveAllBlockGroups() {
+    LOG(INFO) << "Going to remove all block groups from backend storage, fsId: "
+              << fsId_;
+
+    LockGuard lk(mtx_);
+    for (auto& group : allocatedGroups_) {
+        auto err = ClearBlockGroup(group.second);
```

Comment:
mark as TODO here： etcdclient need to implement remove by prefix.

Reply:
fix